src/Chainweb/Utils/RequestLog.hs

```diff
@@ -0,0 +1,138 @@
+import qualified Data.Text.Encoding as T
+
+import GHC.Generics
+
+import Network.HTTP.Types
+import Network.Wai
+
+import Numeric.Natural
+
+import System.Clock
+import System.LogLevel
+
+-- internal modules
+
+import Chainweb.Logger
+import Chainweb.Utils
+
+-- -------------------------------------------------------------------------- --
+-- Request Logger
+
+data RequestLog = RequestLog
+    { _requestLogVersion :: !T.Text
+    , _requestLogMethod :: !T.Text
+    , _requestLogPath :: ![T.Text]
+    , _requestLogIsSecure :: !Bool
+    , _requestLogRemoteHost :: !T.Text
+    , _requestLogQueryString :: !QueryText
+    , _requestLogBodyLength :: !(Maybe Natural)
+    , _requestLogUserAgent :: !(Maybe T.Text)
+    }
+    deriving (Show, Eq, Ord, Generic)
+    deriving anyclass (NFData, ToJSON)
+
+makeLenses ''RequestLog
+
+-- | INVARIANT: this result of this function must not retain pointers to
+-- the original request data that came over the wire.
+--
+logRequest :: Request -> RequestLog
+logRequest req = RequestLog
+    { _requestLogVersion = sshow $ httpVersion req
+    , _requestLogMethod = T.decodeUtf8 $ requestMethod req
+    , _requestLogPath = pathInfo req
+    , _requestLogIsSecure = isSecure req
+    , _requestLogRemoteHost = sshow $ remoteHost req
+    , _requestLogQueryString = queryToQueryText $ queryString req
+    , _requestLogBodyLength = case requestBodyLength req of
+        ChunkedBody -> Nothing
+        KnownLength x -> Just $ int x
+    , _requestLogUserAgent = T.decodeUtf8 <$> requestHeaderUserAgent req
+    }
+
+requestLogger :: Logger l => l -> Middleware
+requestLogger logger app req respond = do
+    logFunctionJson logger Info $ logRequest req
+    app req respond
+
+-- -------------------------------------------------------------------------- --
+-- Request-Response Logger
+
+data RequestResponseLog = RequestResponseLog
+    { _requestResponseLogRequest :: !RequestLog
+    , _requestResponseLogStatus :: !T.Text
+    , _requestResponseLogDurationMicro :: !Int
+    }
+    deriving (Show, Eq, Ord, Generic)
+    deriving anyclass (NFData, ToJSON)
+
+makeLenses ''RequestResponseLog
+
+logRequestResponse :: RequestLog -> Response -> Int -> RequestResponseLog
+logRequestResponse reqLog res d = RequestResponseLog
+    { _requestResponseLogRequest = reqLog
+    , _requestResponseLogStatus = sshow $ responseStatus res
+    , _requestResponseLogDurationMicro = d
+    }
+
+-- | NOTE: this middleware should only be used for APIs that don't stream. Otherwise
+-- the logg may be delayed for indefinite time.
+--
+requestResponseLogger :: Logger l => l -> Middleware
+requestResponseLogger logger app req respond = do
+    let !reqLog = logRequest req
+    reqTime <- getTime Monotonic
+    app req $ \res -> do
+        resTime <- getTime Monotonic
+        logFunctionJson logger Info
+            $ logRequestResponse reqLog res
+            $ (int $ toNanoSecs $ diffTimeSpec resTime reqTime) `div` 1000
+        respond res
```

Comment:
So the time here is just processing time but doesn't cover the amount of time necessary to send the response body. Should you collect another time delta after `respond` finishes?

Reply:
Oops, good catch. I'll fix it.

Usually, one wants both, "time to first byte" and "total time". One would also want to count the response bytes, but that requires some special casing on the different response body types. I don't think we have long-running streaming endpoints or very data heavy endpoints. So, let's keep it simple and add features as we need them.